jax-stuff.md

```diff
@@ -77,51 +78,52 @@ _styles: >
+and computation, and can communicate with explicit collectives.
 
-Correspondingly, JAX provides two APIs for each of these schools: **jit** (`jax.jit`) and **shard\_map** (`jax.experimental.shard_map.shard_map`).
+| Mode | View? | Explicit sharding? | Explicit Collectives? |
+|---|---|---|---|
+| Auto | Global | ❌ | ❌ |
+| Explicit | Global | ✅ | ❌ |
+| Manual | Per-device | ✅ | ✅ |
 
-1. `jax.jit` lets you specify the sharding of the inputs and outputs to a program (via `in_shardings` and `out_shardings`) and infers the rest using the [GSPMD](https://arxiv.org/abs/2105.04663) compiler. While it isn't perfect, it usually does a decent job at automatically scaling your program to any number of chips.
-2. `jax.experimental.shard_map.shard_map` is the more explicit counterpart. You get a device-local view of the program and have to write any communication you want explicitly. Have a sharded array and want the whole thing on each device? Add a `jax.lax.all_gather`. Want to sum an array across your devices? Add a `jax.lax.psum` (an AllReduce). Programming is harder but far less likely to do something you don't want.
-
-<h3 id="jax-jit-the-automatic-parallelism-solution">jax.jit: the automatic parallelism solution</h3>
+<h3 id="auto-sharding-mode">Auto sharding mode</h3>
 
 jax.jit plays two roles inside JAX. As the name suggests, it "just-in-time" compiles a function from Python into bytecode (via XLA/HLO/LLO) so it runs faster. But if the input is sharded or the user specifies an `in_sharding` or `out_sharding`, it also lets XLA distribute the computation across multiple devices and add communication as needed. For example, here's how you could write a sharded matmul using jax.jit:
 
 ```py
 import jax
 import jax.numpy as jnp
-import jax.sharding as shd
-
-P = shd.PartitionSpec
 
 # Running on an TPU v5e 2x2. This assigns names to the two physical axes of the hardware.
 mesh = jax.make_mesh(axis_shapes=(2, 2), axis_names=('X', 'Y'))
 
 # This tells JAX to use this mesh for all operations, so you can just specify the PartitionSpec P.
-shd.set_mesh(mesh)
+jax.set_mesh(mesh)
 
 # We create a matrix W and input activations In sharded across our devices.
-In = jnp.zeros((8, 2048), dtype=jnp.bfloat16, out_sharding=P('X', 'Y'))
-W = jnp.zeros((2048, 8192), dtype=jnp.bfloat16, out_sharding=P('Y', None))
+In = jnp.zeros((8, 2048), dtype=jnp.bfloat16, device=jax.NamedSharding(mesh, jax.P('X', 'Y')))
+W = jnp.zeros((2048, 8192), dtype=jnp.bfloat16, device=jax.NamedSharding(mesh, jax.P('Y', None)))
 
 def matmul_square(In, W):
   return jnp.einsum('bd,df->bf', jnp.square(In), W)
 
 # We can explicitly compile the sharded matmul function here. This adds all the
 # necessary comms (e.g. an AllReduce after the matmul).
-jit_matmul = jax.jit(matmul_square, out_shardings=P('X', None)).lower(In, W).compile()
+jit_matmul = jax.jit(matmul_square, out_shardings=jax.P('X', None)).lower(In, W).compile()
```

Comment:
Why can I do this here but not in jnp.zeros?

Reply:
Cause jit and jnp.zeros are different APIs ;)

There is a reason behind it. I can explain offline.